# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
import datetime
import matplotlib.pyplot as plt

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\Anthony Suherli\\Desktop\\Springboard\\Guided Capstone\\Step 2'

In [3]:
data = pd.read_csv('step3_output.csv')

In [4]:
dropthis = 'Unnamed: 0'

In [5]:
data.head()
data = data.drop(dropthis,axis =1)

In [6]:
data.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  cluster  
0            30.0        2  
1            80.0        1  
2             0.0        1  
3             0.0        1  
4             0.0        1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [7]:
dfo = data.state
data = pd.concat([data.drop('state',axis = 1),pd.get_dummies(dfo)],axis = 1)

In [8]:
data.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3               Donner Ski Ranch         8012            750      0   
4                  June Mountain        10090           2590      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          0     0       1       5  ...             0   
4        0.0          0          2     0       0       4  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = data.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= data.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x) 

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [11]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
x_train, x_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [12]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(x_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [13]:
# Make a variable called y_pred and 
#assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(x_test)

In [14]:
y_pred

array([ 5.12893276e+01,  5.11505337e+01,  5.08521328e+01,  6.43097744e+01,
        5.71380825e+01,  6.79742641e+01,  5.51006069e+01,  4.16742153e+01,
        4.07225246e+01,  6.82220058e+01,  5.57464809e+01,  4.53374233e+01,
        3.51724377e+01,  5.13782558e+01,  4.65081997e+01,  6.02068081e+01,
        4.03228970e+01,  7.04333706e+01,  7.34789715e+01, -5.29825724e+11,
        5.72475185e+01,  4.43579311e+01, -5.29835023e+11, -5.29825671e+11,
        6.31967983e+01,  6.01453762e+01,  5.67694302e+01,  4.82045803e+01,
        6.80839443e+01,  6.79949551e+01,  3.18540249e+01,  5.73271694e+01,
        6.56477489e+01,  7.10634365e+01,  6.16857510e+01,  3.89476528e+01,
        7.28069973e+01,  3.29710903e+01,  5.39886608e+01,  8.03660488e+01,
        5.62372036e+01,  4.08410549e+01])

## Review Model Outcomes — Iterate over additional models as needed

In [15]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [16]:
print(explained_variance_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

-1.0880077244092149e+20
37844914709.125534


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [17]:
print(lm.intercept_)

-9517882683.614237


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [18]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef = pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])


In [19]:
coef = coef.sort_values(by = 'Coefficient',ascending = False)

In [20]:
print(coef)

                    Coefficient
trams              1.483311e+12
fastSixes          1.137365e+12
total_chairs       6.383907e+11
surface            3.646951e+11
double             3.529275e+11
triple             3.011931e+11
quad               1.821257e+11
New York           1.682380e+11
Michigan           1.514900e+11
fastQuads          1.448313e+11
New Hampshire      1.314241e+11
Pennsylvania       1.257090e+11
Wisconsin          1.257090e+11
fastEight          1.202080e+11
Minnesota          1.131514e+11
California         1.131514e+11
Vermont            1.061758e+11
Idaho              9.860640e+10
New Mexico         9.860640e+10
Connecticut        9.029403e+10
Maine              9.029403e+10
Colorado           9.029403e+10
Wyoming            9.029403e+10
Massachusetts      8.101032e+10
Washington         8.101032e+10
Ohio               8.101032e+10
North Carolina     7.037187e+10
West Virginia      7.037187e+10
Illinois           7.037187e+10
Iowa               7.037187e+10
Oregon  

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [21]:
state_cols = list(data.columns)[25:]

In [22]:
data2 = pd.read_csv('step3_output.csv')

In [23]:
no_state = data2.drop(['Unnamed: 0','Name','AdultWeekend','state'], axis = 1)

In [24]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scale2 = preprocessing.StandardScaler().fit(no_state)



In [25]:
no_state.shape

(167, 23)

In [42]:
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled2 =scale2.transform(no_state) 

In [43]:
x_train2, x_test2, y_train2, y_test2 = train_test_split(no_state, y, test_size=0.25, random_state=1)

In [44]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(x_train2,y_train2)

In [45]:
# Make a variable called y_pred and 
#assign it the result of calling predict() on our model variable with parameter X_test
y_pred2 = model2.predict(x_test2)

In [46]:
y_pred2

array([53.27417685, 53.27164566, 46.5946785 , 65.17284382, 60.41944976,
       64.90491855, 49.84236717, 39.55554602, 45.35790905, 64.46919529,
       53.81111077, 41.8806748 , 35.58054832, 46.40816373, 53.66672387,
       59.18988209, 45.85934603, 71.79560611, 76.41062099, 56.25308911,
       58.73351851, 49.61730755, 75.35508646, 60.29717181, 52.47129397,
       64.11079632, 56.5022684 , 51.3319615 , 67.95775803, 60.29844781,
       41.82449766, 60.5052658 , 60.08232061, 65.47875592, 62.18501852,
       31.1759148 , 71.58338995, 34.2462537 , 57.72385043, 77.35564803,
       63.32780444, 50.03875752])

In [48]:
print(explained_variance_score(y_test2,y_pred2))
print(mean_absolute_error(y_test2,y_pred2))

0.5559565165902218
6.852990686240722


In [49]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef2 = pd.DataFrame(abs(lm2.coef_), no_state.columns, columns=['Coefficient'])

In [50]:
coef2 = coef2.sort_values(by = 'Coefficient',ascending = False)

In [51]:
print(coef2)

                    Coefficient
cluster            3.924873e+00
surface            1.082570e+00
triple             9.208312e-01
AdultWeekday       6.960583e-01
total_chairs       3.875508e-01
quad               3.772698e-01
LongestRun_mi      3.378559e-01
double             2.787507e-01
fastQuads          1.067314e-01
daysOpenLastYear   9.431953e-02
Runs               9.146455e-02
projectedDaysOpen  3.011355e-02
TerrainParks       2.683867e-02
NightSkiing_ac     1.842024e-02
averageSnowfall    1.838393e-02
yearsOpen          9.224870e-03
Snow Making_ac     3.265422e-03
vertical_drop      2.436503e-03
summit_elev        1.056583e-03
SkiableTerrain_ac  3.775805e-04
fastEight          2.678413e-15
trams              1.420739e-15
fastSixes          8.326673e-16


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [37]:
data2

Unnamed: 0                               Name       state  summit_elev  \
0             2                   Hilltop Ski Area      Alaska         2090   
1             4                Sunrise Park Resort     Arizona        11100   
2             5      Yosemite Ski & Snowboard Area  California         7800   
3            10                   Donner Ski Ranch  California         8012   
4            12                      June Mountain  California        10090   
..          ...                                ...         ...          ...   
162         323                      Hogadon Basin     Wyoming         8000   
163         326          Sleeping Giant Ski Resort     Wyoming         7428   
164         327                   Snow King Resort     Wyoming         7808   
165         328  Snowy Range Ski & Recreation Area     Wyoming         9663   
166         329                White Pine Ski Area     Wyoming         9500   

     vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0              294      0        0.0          0          0     0  ...   
1             1800      0        0.0          0          1     2  ...   
2              600      0        0.0          0          0     0  ...   
3              750      0        0.0          0          0     0  ...   
4             2590      0        0.0          0          2     0  ...   
..             ...    ...        ...        ...        ...   ...  ...   
162            640      0        0.0          0          0     0  ...   
163            810      0        0.0          0          0     0  ...   
164           1571      0        0.0          0          0     1  ...   
165            990      0        0.0          0          0     0  ...   
166           1100      0        0.0          0          0     0  ...   

     SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0                 30.0       30.000000        150.000000       36.0   
1                800.0       80.000000        115.000000       49.0   
2                 88.0      174.873239        110.000000       84.0   
3                505.0       60.000000        163.000000       82.0   
4               1500.0      174.873239        115.103943       58.0   
..                 ...             ...               ...        ...   
162               92.0       32.000000        121.000000       61.0   
163              184.0       18.000000         61.000000       81.0   
164              400.0      250.000000        121.000000       80.0   
165               75.0       30.000000        131.000000       59.0   
166              370.0      174.873239        115.103943       81.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0               69.0     30.000000      34.00000         152.000000   
1              250.0     74.000000      78.00000         104.000000   
2              300.0     47.000000      47.00000         107.000000   
3              400.0     75.000000      75.00000         170.000000   
4              250.0     57.916957      64.16681         128.000000   
..               ...           ...           ...                ...   
162             80.0     48.000000      48.00000          95.000000   
163            310.0     42.000000      42.00000          77.000000   
164            300.0     59.000000      59.00000         123.000000   
165            250.0     49.000000      49.00000         120.053004   
166            150.0     57.916957      49.00000         120.053004   

     NightSkiing_ac  cluster  
0              30.0        2  
1              80.0        1  
2               0.0        1  
3               0.0        1  
4               0.0        1  
..              ...      ...  
162             0.0        1  
163             0.0        1  
164           110.0        1  
165             0.0        1  
166             0.0        1  

[167 rows x 27 columns]

In [67]:
test3 = data2.drop(['Unnamed: 0','Name','AdultWeekend','state','summit_elev'],axis = 1)

In [68]:
scale3 = preprocessing.StandardScaler().fit(test3)

In [69]:
X_scaled3 =scale3.transform(test3) 

In [70]:
x_train3, x_test3, y_train3, y_test3 = train_test_split(X_scaled3, y, test_size=0.25, random_state=1)

In [71]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(x_train3,y_train3)

In [72]:
y_pred3 = model3.predict(x_test3)

In [73]:
y_pred3

array([54.21167786, 51.49355345, 48.02382481, 65.02689122, 53.31995991,
       62.91192296, 48.19734304, 38.83878075, 44.80462353, 67.22116718,
       54.09134529, 41.57152873, 36.27551813, 46.87635043, 53.5064588 ,
       59.58847825, 51.07740232, 66.43326902, 82.09355907, 56.17909917,
       58.51774593, 49.3169168 , 74.58825821, 58.52193323, 53.19479151,
       62.84882823, 56.55118695, 56.59730844, 66.32988244, 59.97314324,
       41.05342937, 59.3944717 , 58.71464057, 63.96907923, 60.59833958,
       31.37820307, 71.16558495, 32.54755481, 55.84171998, 76.7815593 ,
       64.61482045, 49.72563709])

In [74]:
print(explained_variance_score(y_test3,y_pred3))
print(mean_absolute_error(y_test3,y_pred3))

0.54130802978781
7.275361831392759


In [75]:
coef3 = pd.DataFrame(abs(lm3.coef_), test3.columns, columns=['Coefficient'])

In [76]:
coef3 = coef3.sort_values(by='Coefficient',ascending=False)

In [77]:
coef3

Coefficient
AdultWeekday       1.071837e+01
averageSnowfall    2.856805e+00
cluster            2.234782e+00
vertical_drop      1.905925e+00
SkiableTerrain_ac  1.761634e+00
daysOpenLastYear   1.459799e+00
triple             1.229914e+00
Runs               1.099987e+00
NightSkiing_ac     1.028189e+00
surface            9.014494e-01
double             8.611187e-01
total_chairs       6.204359e-01
quad               2.787251e-01
Snow Making_ac     2.627854e-01
TerrainParks       2.300518e-01
projectedDaysOpen  1.400352e-01
LongestRun_mi      5.530607e-02
yearsOpen          2.753312e-02
fastQuads          1.949934e-03
fastSixes          8.881784e-16
trams              4.440892e-16
fastEight          2.220446e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

In [79]:
print(explained_variance_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

-1.0880077244092149e+20
37844914709.125534


In [80]:
print(explained_variance_score(y_test2,y_pred2))
print(mean_absolute_error(y_test2,y_pred2))

0.5559565165902218
6.852990686240722


In [78]:
print(explained_variance_score(y_test3,y_pred3))
print(mean_absolute_error(y_test3,y_pred3))

0.54130802978781
7.275361831392759


| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -inf|  |-|
| Model 2. | 0.556| |'state'|
| Model 3. | 0.543 | |'state','summit_elev','base_elev'|

Model Selection: